src/download.c

```diff
@@ -306,18 +310,51 @@ static int perform_curl_io_and_complete(struct swupd_curl_parallel_handle *h, in
 	return 0;
 }
 
-/* Fixed window added to hysteresis upper bound before it is enforced. */
-#define BUFFER 5
+static CURLMcode process_curl_multi_wait(CURLM *mcurl, int *repeats)
+{
+	CURLMcode ret;
+	int numfds = 0;
+
+	ret = curl_multi_wait(mcurl, NULL, 0, CURL_MULTI_TIMEOUT, &numfds);
+
+	// If numfds is zero curl_multi_wait reached a timeout or there isn't
+	// any fd to process. The first ocurrence we assume it's a timeout.
+	// After that we assume we need to force a sleep before calling
+	// curl_multi_wait() again.
+	if (ret != CURLM_OK && !numfds) {
+		if (*repeats == 1) {
+			usleep(100000); // sleep 100ms
+		}
+		*repeats = 1;
+	} else {
+		*repeats = 0;
+	}
 
-/* 2 limits, so that we can have hysteresis in behavior. We let the caller
- * add new transfer up until the queue reaches the high threshold. At this point
- * we don't return to the caller and instead process the queue until its len
- * gets below the low threshold */
+	return ret;
+}
+
+/* Poll for downloads if the number of enqueued files are between the lower and
+  * higher limits so we can have hysteresis in behavior.
+  *
+  * If the number of scheduled downloads is lower than xfer_queue_high the
+  * function returns without processing the downloads.
+  *
+  * If the number of scheduled downloads is equal or higher than
+  * xfer_queue_high we will process downloads and wait until the length of the
+  * download queue is lower than xfer_queue_low.
+  *
+  * There's a protection to avoid having a download queue larger than
+  * xfer_queue_high + XFER_QUEUE_BUFFER in cases where curl doesn't have any
+  * data to process for any download. In this case we will return an error and
+  * the caller is in charge of setting the last download to be retried.
+  */
 static int poll_fewer_than(struct swupd_curl_parallel_handle *h, size_t xfer_queue_high, size_t xfer_queue_low)
 {
 	CURLMcode curlm_ret;
-	int running;
+	int running = 1;
```

Comment:
use the available `bool` type please.

Reply:
We can't use bool type for running because it's used by a curl API that requires an int. I'm assuming your comment is about the repeats variable that can be a bool variable. I'll fix that. thanks